rq/worker.py

```diff
@@ -667,24 +669,36 @@ def perform_job(self, job, queue):
                 # to use the same exc handling when pickling fails
                 job._result = rv
 
-                self.set_current_job_id(None, pipeline=pipeline)
-
                 result_ttl = job.get_result_ttl(self.default_result_ttl)
                 if result_ttl != 0:
                     job.ended_at = utcnow()
-                    job.set_status(JobStatus.FINISHED, pipeline=pipeline)
-                    job.save(pipeline=pipeline)
 
-                    finished_job_registry = FinishedJobRegistry(job.origin,
-                                                                self.connection)
-                    finished_job_registry.add(job, result_ttl, pipeline)
+                while True:
+                    try:
+                        # if dependencies are inserted after enqueue_dependents
+                        # a WatchError is thrown by execute()
+                        pipeline.watch(job.dependents_key)
```

Comment:
I'm sorry to interrupt you again, but that doesn't work either, respectively it doesn't fix the race condition. What if the following happens sequentially:

1) Your suggested block is executed, so all current dependencies are enqueued. The job is still not finished, as this should happen in a separate transaction as you suggested.
2) afterwards a new dependency is enqueued (while the worker is running)
3) Now the worker executes the job status/deletion changes in another transaction and sets the job to FINISHED

The result is a deferred job that gets never executed.
We have to enqueue the jobs AND change the status in ONE atomic transaction, to avoid a race condition. Also important related to redis is, that if a `WatchError` occurs, you have to replay all statements related to the transaction (normally starting with a `watch`) again.

So I think we could only extract the code and maybe find some other changes (eg. move all the code from `enqueue_dependents` into `handle_successful_job_execution`) to improve the readability.